### PR TITLE
SNOW-777893: improve type hint of SnowflakeCursor.execute

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-777893

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/connector/cursor.py [hunk 1]
```diff
@@ -607,6 +607,66 @@ def _preprocess_pyformat_query(
...
+        _put_azure_callback: SnowflakeProgressPercentage = None,
+        _put_callback_output_stream: IO[str] = sys.stdout,
+        _get_callback: SnowflakeProgressPercentage = None,
+        _get_azure_callback: SnowflakeProgressPercentage = None,
+        _get_callback_output_stream: IO[str] = sys.stdout,
+        _show_progress_bar: bool = True,
+        _statement_params: dict[str, str] | None = None,
+        _is_internal: bool = False,
+        _describe_only: bool = False,
+        _no_results: Literal[False] = False,
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/cursor.py:630]: Why use `Literal[False]` instead of `bool`?

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/connector/cursor.py [hunk 2]
```diff
@@ -607,6 +607,66 @@ def _preprocess_pyformat_query(
...
+        _put_azure_callback: SnowflakeProgressPercentage = None,
+        _put_callback_output_stream: IO[str] = sys.stdout,
+        _get_callback: SnowflakeProgressPercentage = None,
+        _get_azure_callback: SnowflakeProgressPercentage = None,
+        _get_callback_output_stream: IO[str] = sys.stdout,
+        _show_progress_bar: bool = True,
+        _statement_params: dict[str, str] | None = None,
+        _is_internal: bool = False,
+        _describe_only: bool = False,
+        _no_results: Literal[True] = True,
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/cursor.py:660]: Ditto.

File: src/snowflake/connector/cursor.py (hunk 1 above)
Comment (sfc-gh-sfan) [reply src/snowflake/connector/cursor.py:630]: I think probably because the value of `_no_results` determines the actual return type of this function?

Comment (sfc-gh-aling) [reply src/snowflake/connector/cursor.py:630]: shixuan hit the nail, it overrides the signature to make the type hint clear on the return type depending on the `_no_results`

**Pull request merged**